contracts/MerkleDistributor.sol

```diff
@@ -0,0 +1,115 @@
+        uint256 claimedWord = claimedBitMap[claimedWordIndex];
+        uint256 mask = (1 << claimedBitIndex);
+        return claimedWord & mask == mask;
+    }
+
+    function _setClaimed(uint256 index) private {
+        uint256 claimedWordIndex = index / 256;
+        uint256 claimedBitIndex = index % 256;
+        claimedBitMap[claimedWordIndex] = claimedBitMap[claimedWordIndex] | (1 << claimedBitIndex);
+    }
+
+
+    function claim(bytes32 tokenSymbol, uint256 tokenIndex, uint256 amount, bytes calldata prefixNode, bytes calldata suffixNode, bytes calldata ownerSignature, bytes calldata approvalSignature, bytes32[] calldata merkleProof) external override {
+        ClaimInfo memory claimInfo;
+        claimInfo.tokenHash = keccak256(abi.encodePacked(tokenIndex, tokenSymbol, amount));
+        claimInfo.node = keccak256(abi.encodePacked(prefixNode, claimInfo.tokenHash, suffixNode));
+        // TODO: consider the better way to index the claim info.
+        claimInfo.index = keccak256(tokenSymbol, bytes32(tokenIndex), claimInfo.node);
+    
+        // Check if the token is claimed.
+        require(isClaimed(claimInfo.index), "AlreadyClaimed");
+
+        // Check if the token is exist.
+        claimInfo.contractAddr = _checkTokenContractExist(tokenSymbol);
+
+        // Verify the approval signature.
+        _verifySignature(tokenSymbol, msg.sender, ownerSignature, approvalSignature, claimInfo.node);
+    
+        // Verify the merkle proof.
+        require(MerkleProof.verify(merkleProof, merkleRoot, claimInfo.node), "InvalidProof");
+
+        // Check balance of the contract. make sure Tokenhub has enough balance.
+        require(IBEP20(claimInfo.contractAddr).balanceOf(TOKEN_HUB_ADDR) >= amount, "InsufficientBalance");
+
+        // Mark it claimed and send the token.
+        _setClaimed(claimInfo.index);
+    
+        // Unlock the token from TokenHub.
+        ITokenHub(TOKEN_HUB_ADDR).unlock(claimInfo.contractAddr, msg.sender, amount);
+
+        emit Claimed(tokenSymbol, msg.sender, amount);
+    }
+
+    function registerToken(bytes32 tokenSymbol, address contractAddr, uint256 decimals, uint256 amount, bytes calldata ownerSignature, bytes calldata approvalSignature) external override {
+         // Check if the token is exist.
+        _checkTokenContractExist(tokenSymbol);
+
+        // Verify the approval signature.
+        _verifySignature(tokenSymbol, msg.sender, ownerSignature, approvalSignature, bytes32(amount));
+
+        // Check balance of the contract. make sure Tokenhub has enough balance.
+        require(IBEP20(contractAddr).balanceOf(TOKEN_HUB_ADDR) >= amount, "InsufficientBalance");
+
+        // Bind the token to TokenHub.
+        ITokenHub(TOKEN_HUB_ADDR).bindToken(tokenSymbol, contractAddr, decimals);
+    }
+
+    function _verifySignature(bytes32 tokenSymbol, address account, bytes memory ownerSignature, bytes memory approvalSignature, bytes32 node) private pure {
```

Comment:
We need to verify `ownerSignature` is signed by the token owner, dont we?

Reply:
yes, I will implement the precompile contract later